dashboard/app/helpers/javalab_files_helper.rb

```diff
@@ -15,53 +14,47 @@ def self.upload_project_files(channel_id, level_id, hostname, auth_token)
+  # This method adds in user-specific code and assets uploaded on the level where the project was created.
+  # The returned hash is in the format below. All values are strings.
   # {
   #   "sources": {"main.json": <main source file for a project>, "grid.txt": <serialized maze if it exists>},
   #   "assetUrls": {"asset_name_1": <asset_url>, ...}
   #   "validation": <all validation code for a project, in json format>
   # }
-  # If the channel doesn't have validation and/or a maze, those fields will not be present.
+  # If the level doesn't have validation and/or a maze, those fields will not be present.
   def self.get_project_files(channel_id, level_id)
-    all_files = {}
-    sources = {}
+    all_files = get_level_files(level_id)
+
     # get main.json
     source_data = SourceBucket.new.get(channel_id, "main.json")
-    sources["main.json"] = source_data[:body].string
-
-    # get maze file
-    level = Level.find(level_id)
-    if level
-      serialized_maze = level.try(:get_serialized_maze)
-      if serialized_maze
-        sources["grid.txt"] = serialized_maze.to_json
-      end
-    end
-    all_files["sources"] = sources
+    all_files["sources"]["main.json"] = source_data[:body].string
 
     # get level assets
-    assets = {}
     asset_bucket = AssetBucket.new
     asset_list = asset_bucket.list(channel_id)
     asset_list.each do |asset|
-      assets[asset[:filename]] = generate_asset_url(asset[:filename], channel_id)
-    end
-
-    # get starter assets
-    if level
-      (level&.project_template_level&.starter_assets || level.starter_assets || []).map do |friendly_name, _|
-        assets[friendly_name] = generate_starter_asset_url(friendly_name, level)
-      end
+      all_files["assetUrls"][asset[:filename]] = generate_asset_url(asset[:filename], channel_id)
     end
-    all_files["assetUrls"] = assets
 
-    # get validation code
-    if level.respond_to?(:validation) && level.validation
-      all_files["validation"] = level.validation.to_json
-    end
+    all_files
+  end
 
-    return all_files
+  # Get all files for the project to be executed as a hash, with source code provided as an argument.
+  # Much of this can be constructed from the level where this project was created (get_level_files).
+  # This method adds in code provided the sources argument.
+  # The returned hash is in the format below. All values are strings.
+  # {
+  #   "sources": {"main.json": <main source file for a project>, "grid.txt": <serialized maze if it exists>},
+  #   "assetUrls": {"asset_name_1": <asset_url>, ...}
+  #   "validation": <all validation code for a project, in json format>
+  # }
+  # If the level doesn't have validation and/or a maze, those fields will not be present.
+  def self.get_project_files_with_override_sources(sources, level_id)
+    all_files = get_level_files(level_id)
+    all_files["sources"]["main.json"]["source"] = sources
```

Comment:
It's possible this will change _slightly_ once implemented on the javalab side, but this is the file structure I see when I look at sources in S3.